app/models/king.rb

```diff
@@ -1,7 +1,51 @@
 class King < Piece
+  def valid_move?(to_row, to_col)
+    logger.debug "*** KING: #{id} #{type} valid_move?"
+    return false if move_nil?(to_row, to_col)
+    return false if move_out_of_bounds?(to_row, to_col)
+    return false if move_destination_ally?(to_row, to_col)
+    return false unless move_legal?(to_row, to_col)
+    true
+  end
+
   def move_legal?(to_row, to_col)
+    logger.debug "move_legal for #{type}"
     row_diff = (row - to_row).abs
     col_diff = (col - to_col).abs
-    row_diff <= 1 && col_diff <= 1
+    return true if row_diff <= 1 && col_diff <= 1
+
+    legal_castle_move(to_row, to_col)
+  end
+
+  private
+
+  def legal_castle_move(to_row, to_col)
+    return false unless legal_king_move?(to_row, to_col)
+    legal_rook_move?(to_col)
+  end
+
+  def legal_king_move?(to_row, to_col)
+    return false unless moves.empty?
+    return false unless row == to_row
+    return false unless (col - to_col).abs == 2
+    true
+  end
+
+  def legal_rook_move?(to_col)
+    rook = get_rook?(to_col)
+    return false unless rook
+    return false if move_obstructed?(row, rook.col)
+    new_rook_col = rook.col.zero? ? 3 : 5
+    rook.update_piece(row, new_rook_col, 'castled')
```

Comment:
This `update_piece` should probably be moved into its own method so that the check methods (the ones ending in `?`) just answer questions about *if* it's okay, and then a method like `castle!(to_col)` actually does the castling.

```ruby
# NOTE: the king will be moved through the normal move_to!
def castle!(to_col)
  rook = get_castling_rook(to_col)
  return unless rook
  new_rook_col = rook.col.zero? ? 3 : 5
  rook.update_piece(row, new_rook_col, 'castled')
end
```

You could add a method named `is_castling?(to_row, to_col)` to the Piece class:

```ruby
def is_castling?(to_row, to_col)
  false
end
```

Then override it in the King class:

```ruby
def is_castling?(to_row, to_col)
  row == to_row && (col - to_col).abs == 2
end
```

Then inside `move_to!` in the Piece class, right after `capture_piece`, you could have:

```ruby
# ...
capture_piece(to_row, to_col)
castle!(to_col) if is_castling?(to_row, to_col)
# ...
```

That would take care of moving the rook.

Just one idea for how to separate out the validation from the actual move.